### PR TITLE
Implement CLI options to set truncation thresholds

### LINKED ISSUE #12765: Customisable truncation limits
Hi! 
I've recently encountered a problem regarding basic truncation limits.

### What's the problem this feature will solve?
[Truncation method](https://github.com/pytest-dev/pytest/blob/main/src/_pytest/assertion/truncate.py#L36-L37) allows to set custom `max_lines`, `max_chars`, but this mechanism is not used in its only call [here](https://github.com/pytest-dev/pytest/blob/main/src/_pytest/assertion/truncate.py#L24), even when `max_length` (I guess it is used as a synonym to `max_chars`) can be specified in `truncate_if_required`.
However, `truncate_if_required` is not called with any length parameter (see [this invocation](https://github.com/pytest-dev/pytest/blob/main/src/_pytest/assertion/__init__.py#L156)).

In its basic form, it produces quite small diff, that may not be as intelligible as it is expected to be :(

#### Example
test.py file: 
```py
string_a = """
Code sample:
def a():
    print(10000) #comment
    do_something()

def b():
    print(100000) #comment
    do_something()
"""

string_b = """
Code sample:
def a():
    print(10000) # comment
    do_something()

def b():
    print(100000) # comment
    do_something()
"""

def test():
    assert string_a == string_b

```

This results in 

```
% pytest test.py
============================================================================================== test session starts ==============================================================================================
platform darwin -- Python 3.12.3, pytest-8.3.2, pluggy-1.5.0
rootdir: <root-dir>
collected 1 item                                                                                                                                                                                                

test.py F                                                                                                                                                                                                 [100%]

=================================================================================================== FAILURES ====================================================================================================
_____________________________________________________________________________________________________ test ______________________________________________________________________________________________________

    def test():
>       assert string_a == string_b
E       AssertionError: assert '\nCode sampl...something()\n' == '\nCode sampl...something()\n'
E         
E           
E           Code sample:
E           def a():
E         -     print(10000) # comment
E         ?                   -
E         +     print(10000) #comment...
E         
E         ...Full output truncated (7 lines hidden), use '-vv' to show

test.py:26: AssertionError
============================================================================================ short test summary info ============================================================================================
FAILED test.py::test - AssertionError: assert '\nCode sampl...something()\n' == '\nCode sampl...something()\n'
=============================================================================================== 1 failed in 0.01s ===============================================================================================
                                                                                                      
```
As seen here, second diff is hidden, when it appears in full diff (with `-vv` flag specified) (see below):

```
    def test():
>       assert string_a == string_b
E       AssertionError: assert '\nCode sample:\ndef a():\n    print(10000) #comment\n    do_something()\n\ndef b():\n    print(100000) #comment\n    do_something()\n' == '\nCode sample:\ndef a():\n    print(10000) # comment\n    do_something()\n\ndef b():\n    print(100000) # comment\n    do_something()\n'
E         
E           
E           Code sample:
E           def a():
E         -     print(10000) # comment
E         ?                   -
E         +     print(10000) #comment
E               do_something()
E           
E           def b():
E         -     print(100000) # comment
E         ?                    -
E         +     print(100000) #comment
E               do_something()

test.py:24: AssertionError
``` 

### Describe the solution you'd like

Some solutions are as follows: 
1. Increase the threshold of truncation (e.g. to 16 lines and respectively, to 16 * 80 chars), but this is still not customisable.
2. Make those limits customizable (for example, with CLI flags (`--truncation-limit-lines` and `--truncation-limit-chars` or something like that). 

### Additional solutions
This may be surpassed by adding `-vv` flag, however, there are some problems with this mode, for example, #8998.

### PR DESCRIPTION
**Pull request by zhukoff-pavel**
closes #12765 

Example implementation of truncation limit setting with CLI flags

### CODE REVIEW

**Comment by nicoddemus** [comment]
Thanks @zhukoff-pavel, we appreciate the PR!

I would like to see what other maintainers think of this feature, overall I'm +1 on it, the only downside I see is the cognitive load of extra options...   but perhaps others have different opinions here.

I also wonder if this is better implemented as an ini option instead... seems like it would be more common to set this in the config file in test suites where this feature is often wanted (and of course it is always pass this using the `-o` option in the command line).

File: changelog/12765.feature.rst
Comment (nicoddemus) [inline changelog/12765.feature.rst:1]: Nice 👍 

Perhaps we should also update [`output.rst`](doc/en/how-to/output.rst) to mention the new command-line options (or ini options as per my comment).

**Comment by zhukoff-pavel** [comment]
@nicoddemus, thank you for your reply! 

> I also wonder if this is better implemented as an ini option instead... seems like it would be more common to set this in the config file in test suites where this feature is often wanted (and of course it is always pass this using the `-o` option in the command line).

Maybe it is better to implement an `.ini` option *and* CLI flag to e.g. temporarily override limits in a singular launch? 

Is it possible to implement such behaviour with current option reading machinery (with some combination of `addoption` and `addini` methods)?

**Comment by nicoddemus** [comment]
> Maybe it is better to implement an .ini option and CLI flag to e.g. temporarily override limits in a singular launch?

Your suggestion makes sense, however we have since then moved away from this because:

1. It involves a lot of code duplication
2. We have mechanisms today where users can use them interchangeably (use `addopts` to hard-code command-line options in the ini file, and use `-o` to pass config options in the command line).

For this reason nowadays we prefer to have either a command-line option (for features where users are likely to change between pytest invocations) or ini option (where users are most likely to set once per repository).

> Is it possible to implement such behaviour with current option reading machinery (with some combination of addoption and addini methods)?

Unfortunately no, it requires duplicating the code.

**Comment by zhukoff-pavel** [comment]
I see.
I can then rewrite this PR to implement .ini option instead, as it definitely looks more like a one-time config setting rather than per launch.

**Comment by nicoddemus** [comment]
> I can then rewrite this PR to implement .ini option instead

Thanks! I suggest to wait a bit in case some other maintainer has comments or reservations about the feature though.

**Comment by nicoddemus** [comment]
Been awhile and nobody seems to be against the idea, so feel free to go ahead.

File: changelog/12765.feature.rst
Comment (nicoddemus) [inline changelog/12765.feature.rst:1]:
```suggestion
Thresholds to trigger snippet truncation can now be set with :confval:`truncation_limit_lines` and :confval:`truncation_limit_chars`.
```

File: doc/en/how-to/output.rst
Comment (nicoddemus) [inline doc/en/how-to/output.rst:565]:
```suggestion
That will cause pytest to truncate the assertions to 10 lines or 90 characters, whichever comes first.
```

File: src/_pytest/assertion/truncate.py
Comment (nicoddemus) [inline src/_pytest/assertion/truncate.py:23]: Do we want to disable truncation altogether if users set `truncation_limit_lines` to `0`?

I believe so, in which case we need to handle `None` explicitly here, as `0` will short circuit to `DEFAULT_MAX_LINES` (same applies to the code below).

File: testing/test_assertion.py
Comment (nicoddemus) [inline testing/test_assertion.py:1444]: Let's add a test case for no truncation at all.

**Review by nicoddemus — CHANGES_REQUESTED** [review]
Thanks @zhukoff-pavel!

Please take a look at my comments.

File: testing/test_assertion.py
Comment (zhukoff-pavel) [reply testing/test_assertion.py:1444]: Is that a test case, when the limits are too high to perform any truncation? If so, that is done in test-cases in lines `1442` and `1444`. 

Or is it something completely different and more applicable to the case stated in your message [here](https://github.com/pytest-dev/pytest/pull/12766#discussion_r1761602377)?

File: src/_pytest/assertion/truncate.py
Comment (zhukoff-pavel) [reply src/_pytest/assertion/truncate.py:23]: I guess, that may be better to implement this logic as an additional check to `_should_truncate_item` method?

Comment (zhukoff-pavel) [reply src/_pytest/assertion/truncate.py:23]: Do we want to disable truncation if and only if both conditions (`truncation_limit_lines` and `truncation_limit_chars` are set to `0`) are met?

File: testing/test_assertion.py
Comment (nicoddemus) [reply testing/test_assertion.py:1444]: Ahh sorry, I should have been clearer: I meant a case where the user explicitly sets both options to `0`, and then the output is not truncated at all.

File: src/_pytest/assertion/truncate.py
Comment (nicoddemus) [reply src/_pytest/assertion/truncate.py:23]: Ahh yes, makes more sense to have this logic in `_should_truncate_item`.

Comment (zhukoff-pavel) [reply src/_pytest/assertion/truncate.py:23]: However, we still do need to have an explicit `0` handling, not here, but further down the stack (Here `0` is interpreted as a string and thus will be passed further into function, being transformed into `int` value), because in the current situation full truncations are possible by setting one of the values to `0`.

I think of a scenario as follows:
If either of `truncation_limit_lines` and `truncation_limit_chars` are set to `0` (but not both, that will be caught by `_should truncate_item`), we disable that exact truncation, but try to truncate with the other method, to avoid unwanted behaviour.

So, if the `truncation_limit_lines` is set to `0`, we do not try to truncate lines, but we try to truncate by `truncation_limit_chars`, and vice versa. 

That will look something like this
```diff
@@ -79,10 +88,11 @@ def _truncate_explanation(
     ):
         return input_lines
     # Truncate first to max_lines, and then truncate to max_chars if necessary
-    truncated_explanation = input_lines[:max_lines]
+    if max_lines > 0:
+        truncated_explanation = input_lines[:max_lines]
     truncated_char = True
     # We reevaluate the need to truncate chars following removal of some lines
-    if len("".join(truncated_explanation)) > tolerable_max_chars:
+    if len("".join(truncated_explanation)) > tolerable_max_chars and max_chars > 0:
```

**Comment by zhukoff-pavel** [comment]
@nicoddemus , thank you for a thorough review! 

Please have a look at the changes. 

I implemented some logic regarding value of `0` being set in one mode or both at a time.

File: src/_pytest/assertion/truncate.py
Comment (zhukoff-pavel) [reply src/_pytest/assertion/truncate.py:39]: Yes, I also did not like it when I implemented this, sounds good to refactor this place :) 

Will think, what I can do here

However, I do think about return values here. `None` sounds great, but will it be clear that it means "No truncation" instead of using default values?

Comment (nicoddemus) [reply src/_pytest/assertion/truncate.py:39]: > None sounds great, but will it be clear that it means "No truncation" instead of using default values?


You got a point, for a more visible API probably a new type would be best, but given this is internal and used in a single place, I guess using `None` is fine (as long as if documented of course).

Comment (zhukoff-pavel) [reply src/_pytest/assertion/truncate.py]: I've decided to return an additional `should_truncate` value. Seems simpler not to derive the truncation status somewhere else.

File: doc/en/how-to/output.rst
Comment (nicoddemus) [inline doc/en/how-to/output.rst:565]: Ahh sorry I forgot, we also need to explicitly document these options in [reference.rst](https://github.com/pytest-dev/pytest/blob/main/doc/en/reference/reference.rst)... see the `confval` definitions at the bottom. Please add one section for each of these options, in alphabetical order. :+1:

File: src/_pytest/assertion/truncate.py
Comment (nicoddemus) [inline src/_pytest/assertion/truncate.py:32]: We should update the docstring:

```suggestion
    """Return the truncation parameters related to the given item, as (should truncate, max lines, max chars)."""
```

**Review by nicoddemus — CHANGES_REQUESTED** [review]
Thanks @zhukoff-pavel!

Two minor things still missing, but after that it LGTM! :grin: 

Thanks for the patience.

File: doc/en/how-to/output.rst
Comment (nicoddemus) [reply doc/en/how-to/output.rst:565]: (that's why the readthedocs build failed)

File: doc/en/how-to/output.rst
Comment (nicoddemus) [inline doc/en/how-to/output.rst:553]:
```
doc/en/how-to/output.rst:554: WARNING: Explicit markup ends without a blank line; unexpected unindent. [docutils]
```

```suggestion
.. _truncation-params:

```

**Comment by zhukoff-pavel** [comment]
@nicoddemus, thank you for review and ideas!

File: doc/en/reference/reference.rst
Comment (nicoddemus) [inline doc/en/reference/reference.rst:1891]:
```suggestion
   .. note::
```

File: doc/en/reference/reference.rst
Comment (nicoddemus) [inline doc/en/reference/reference.rst:1911]:
```suggestion
   .. note::
```

**Review by nicoddemus — APPROVED** [review]
Thanks a lot @zhukoff-pavel, appreciate the patience! :+1: 

Leaving this open for awhile in case anybody else wants to review it, otherwise I will merge it soon.

**Pull request merged**
